### ISSUE TITLE
清除梯度的操作是否必要？

### ISSUE THREAD

**Issue by shuguang99** [Outsider565/LoRA-GA#17]
for n, p in model.named_parameters():
            if p.grad is not None:
                p.grad = None
请问这段代码的操作是必要的吗，在我的运行过程中，p.grad = None似乎不会被执行

**Comment by shuguang99** [comment]
get_record_gradient_hook(model, named_grads)(None)  # get gradient of last layer
同样的，请问这段代码是必要的吗，我没太理解他的作用～

**Comment by shuguang99** [comment]
谢谢作者的工作！！！真的很棒

**Comment by zeroDtree** [comment]
- get_record_gradient_hook 是必须的

因为在`/peft/utils/lora_ga_utils/offload_utils_for_quant/gradient_offload.py`这个文件里的`GradientOffloadHookContext`类里是通过`torch.Tensor.register_hook`来 hook 梯度的。

```python
    def register_gradient_hook(self):
        for _, param in self.model.named_parameters():
            hook = param.register_hook(
                self.get_record_gradient_hook(self.model, self.record_dict)
            )
            self.handle_list.append(hook)

    def get_record_gradient_hook(self, model, record_dict):
        def record_gradient_hook(grad):
            for n, p in model.named_parameters():
                if p.requires_grad and p.grad is not None:
                    if n not in record_dict:
                        record_dict[n] = p.grad.to(self.gradient_device)
                    else:
                        record_dict[n] += p.grad.to(self.gradient_device)
                    p.grad = None
            return grad

        return record_gradient_hook
```

这个`record_gradient_hook`的返回值会成为 tensor 的最终得到的梯度。

`record_gradient_hook(grad)`只是把每次得到的 tensor 的梯度累加到`record_dict`里，然后清除 tensor 的梯度。

`record_gradient_hook(grad)`的工作方式不是同步的，当前结点的梯度在之后的反向传播中被累加到`record_dict`里。

所以最后一个被反向传播到的结点的得到梯度后，需要手动执行一下这个 hook,因为对最后一个结点来说，没有比它更晚被反向传播的结点了。

torch.Tensor.register_hook 的文档：[torch.Tensor.register_hook](https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html)

- 这段代码不是不必须的

```
    for n, p in model.named_parameters():
        if p.grad is not None:
            p.grad = None
```

这段代码是以防万一，防止有 tensor 的梯度没有被清空。

**Comment by shuguang99** [comment]
了解了，感谢！！
